### ISSUE TITLE
empty! should preserve columns

### ISSUE THREAD

**Issue by nalimilan** [JuliaData/DataFrames.jl#1841]
Currently `empty!` removes all columns, making its argument a `0×0 DataFrame`. This doesn't seem consistent with the view of data frames as collections of rows (https://github.com/JuliaData/DataFrames.jl/issues/1514): according to that definition, `empty!` should just resize all column vectors to have length 0.

With this change, `push!` and `append!` would require columns to match, contrary to what happens for a `0×0 DataFrame`. That can be both a drawback (inconvenient) and an advantage (safety). Another point is that I'm not sure it's very useful to be able to make a `DataFrame` be `0×0`, since one can as well use `DataFrame()`.

We could also just remove that function for now, as it doesn't sound terrible useful.

**Comment by bkamins** [comment]
I was thinking how `isempty` should work earlier and came to the conclusion that its current behavior is OK. Therefore `empty!` should do what you say. Essentially it should be:

```
empty!(df::DataFrame) = empty!.(eachcol(df))
```

> With this change, `push!` and `append!` would require columns to match

They would not be affected. In `push!` and `append!` currently if a data frame has some columns we expect the column names match even if it has 0 rows. Only `DataFrame` with `0` columns accepts adding columns via `push!` and `append!`. (internally we do not test `isempty(df)` but `ncol(df)==0`). Here is an example:

```
julia> df = DataFrame(a=[], b=[])
0×2 DataFrame


julia> push!(df, (x=1, y=2))
ERROR: ArgumentError: Error adding value to column :a.
Stacktrace:
 [1] #push!#132(::Symbol, ::Function, ::DataFrame, ::NamedTuple{(:x, :y),Tuple{Int64,Int64}}) at 
julia> push!(df, (a=1, b=2))
1×2 DataFrame
│ Row │ a   │ b   │
│     │ Any │ Any │
├─────┼─────┼─────┤
│ 1   │ 1   │ 2   │
```

We could remove `empty!` for `Index` though as it is not useful after we change `empty!` for `DataFrame`.

In short - please make a PR if you will as I think it is consistent and not problematic.

**Comment by bkamins** [comment]
Additionally - if you will make this PR (if you prefer me to implement the PR please let me know) can you please fix `select!` so that it accepts empty collections. Now it has a bug that it does `indmin, indmax = extrema(inds)` which assumes non empty collection.

If we fix this bug then to remove all columns from a data frame one can write `select!(df, 0:1)` or something similar.

**Comment by nalimilan** [comment]
> They would not be affected. In `push!` and `append!` currently if a data frame has some columns we expect the column names match even if it has 0 rows. Only `DataFrame` with `0` columns accepts adding columns via `push!` and `append!`. (internally we do not test `isempty(df)` but `ncol(df)==0`).

Precisely, that's what I meant. :-) With the current behavior of `empty!`, you can `push!` any kind of row. If we change it to preserve columns, the names will have to match and types will have to be compatible. Not a big issue, but it's worth noting.

I can make a PR, but we probably need to go through a deprecation phase first? Or do we consider this to be a minor change?

**Comment by bkamins** [comment]
I think we should go through deprecation.

Regarding changes in `select!` please leave them to me. I have just started to implement `Not` for indexing and I need to test it there anyway so I will fix it there. OK?

**Comment by bkamins** [comment]
> Precisely, that's what I meant. :-)

OK - but I do not think anyone uses `empty!` for that (but that is why I think we should go through deprecation)
